[RAM] Add query conditional action filter to Security Solution UI

Zacqary opened the pull request:
## Summary

Closes #152611 
Fixes #155146 too

<img width="788" alt="Screenshot 2023-04-10 at 2 09 29 PM" src="https://user-images.githubusercontent.com/1445834/230977273-9d00aa3d-af2c-4dff-80ef-3034213df90d.png">


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))

Zacqary committed:
Add action alerts filter query

Zacqary committed:
Add submitOnBlur and option overrides to AlertsSearchBar

Zacqary committed:
Add filters to data model

Zacqary committed:
Incorporate filters into generated action values

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

Zacqary committed:
Fix typecheck

Zacqary committed:
Typecheck fix

Zacqary committed:
Fix io-ts

Zacqary committed:
Fix DSL filter tests

Zacqary committed:
io-ts fix

Zacqary committed:
Fix jest

Zacqary committed:
Fix io-ts

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

CoenWarmer reviewed (approved):
AO Changes LGTM

JiaweiWu reviewed (commented):
The code looks good, I saw a few inconsistencies between this query filter and the 1 in discovery:

When trying to use DSLs: 
Discovery: 
![image](https://user-images.githubusercontent.com/74562234/232835461-3fa57024-e42a-4245-ba0c-0cd191498610.png)

New alert filter query:
![image](https://user-images.githubusercontent.com/74562234/232836144-5badbc0e-0c37-4849-a9e2-2a011abec92d.png)

The new alert filter query also seems to lack spacing, and the left button seems to be smaller, just small UI things. 

I will test the filter with Xavier after he shows me how to.

JiaweiWu reviewed (approved):
LGTM, aside from the comment I left earlier :slightly_smiling_face:

Zacqary commented:
The Discover plugin seems to be processing the query DSL and adding it to the filter's `meta.value` prop. Unclear how or where it's doing this.

It's also unclear where the Discover plugin is actually implementing the `<SearchBar>` component so I'm not able to find what this plugin is doing differently.

@elastic/kibana-data-discovery please advise? What do we need to be passing into the `<SearchBar>` to get DSL filter badges to display correctly?

XavierM commented:
@yctercero, I think everything will start to work when @e40pud merged his PR.

ersin-erdal reviewed (commented):
I tested locally and think found a bug.
I created a rule with alerts filter and saved, then edited the rule and removed all the filters (switched `Send alert notification only if alert fields match a query` and. `Send alert notification within the selected time frame only` off)
and got validation error response on update.
i saw we send `alerts_filter: {query: null, timeframe: null}` in the payload.
We should send nothing when those switches are off.

ersin-erdal committed:
make alerts filter query and timeframe optional.
don't send alertsFilter if the both query and timeframe are missing.
use ALERT_LAST_DETECTED instead of kibana.alert.start

ersin-erdal committed:
Merge branch 'main' into query-152611

ersin-erdal committed:
fix failing tests and check types
use @timestamp instead of kibana.alert.last_detected

ersin-erdal committed:
Merge remote-tracking branch 'Zacqary/query-152611' into query-152611

ersin-erdal committed:
fix timeframe in RuleActionAlertsFilter

spong commented:
If saving after first enabling `within selected time frame` I'm seeing an error that says only `00:00-24:00` is supported, however re-selecting `00:00` as the end time seems to get it to pass validation.

Note: a little interesting here that it says `24:00` is a valid value but it isn't listed in the combobox, and when manually entered it will auto-convert back to `00:00`.

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/233483111-971d2396-42e7-40f5-88fd-b20e02f9a93f.gif" />
</p>

ersin-erdal committed:
filter security alerts as well

ersin-erdal committed:
set query on change

ersin-erdal committed:
Merge branch 'main' into query-152611

ersin-erdal commented:
> Hey! This is a really cool feature! Tested it out and am seeing some strange behaviors. Let me know if you need more details from me but I noticed the following:
> 
> * First time going through I created a rule with an action, the query text did not end up saving (didn't get on video)
> * When I edit the query, it sometimes saves, sometimes doesn't (see in video, the edit did not take)
> * When I select to edit my filter, I don't see my values, but if I view it as DSL, I see them (see in video)
> * For one of my filter values it just says "Warning", not sure what to do with that (see in video)
> 
>  154680.mov 
> I did see some React errors like `Can't perform a React state update on an unmounted component.` for `QueryStringInputUI` and `QueryBarMenuComponent` not sure if that's related to what I'm seeing in the video.

Fixed

ersin-erdal commented:
> Testing locally I'm not seeing the `alert_query` functionality working at the moment. I have the main rule query as `host.name:*`, and the `alert_query` as a filter w/ `host.name: exists` (tried `host.name:*` in the query as well), however I'm not seeing my action fired. Note: alerts are being generated, and they do have `host.name` set. Once I disable the `alert_query` I do see the action firing again.
> 

> edit: Also not seeing the `within the selected time frame` option working either. I have it configured for all day Thursday (today) local time, but am only seeing alerts created, not my slack action firing:
> 
Hi @spong 
Must be fixed now, but i've just realised that Security Rules always trigger just one action no matter how many alerts there are.
Therefore, in order to skip an action, your query must filter out all of the alerts.

spong commented:
> Hi @spong
> Must be fixed now, but i've just realised that Security Rules always trigger just one action no matter how many alerts there are.
> Therefore, in order to skip an action, your query must filter out all of the alerts.

Thanks @ersin-erdal! I'll test again later today. That said, since the rule query was the same as the alert query (`host.name:*`) all generated alerts should be matched.

If this is still an issue when I test later I'll provide detailed reproduction steps 👍

ersin-erdal committed:
fix 24:00 -> 00:00 issue

ersin-erdal committed:
Merge remote-tracking branch 'Zacqary/query-152611' into query-152611

ersin-erdal commented:
> If saving after first enabling `within selected time frame` I'm seeing an error that says only `00:00-24:00` is supported, however re-selecting `00:00` as the end time seems to get it to pass validation.
> 
> Note: a little interesting here that it says `24:00` is a valid value but it isn't listed in the combobox, and when manually entered it will auto-convert back to `00:00`.
> 
@yctercero 
It looks like EuiDatePicker doesn't allow us to use 24:00, replaces it with 00:00.
Even though the default time range is  `00:00 -> 24:00` , the component shows it as `00:00 -> 00:00` but posts as `00:00 -> 24:00`.

I changed default end time as 23:59 to avoid confusion. 
And updated code to cover the full 24h when end and start times are the same eg. `00:00 -> 00:00`

ersin-erdal committed:
fix 24:00 -> 00:00 issue

stratoula reviewed (approved):
Unified search changes LGTM, code review only

markov00 reviewed (approved):
LGTM, code review only. The added optional prop doesn't change the current behaviour of the unified search so it is safe on our end.

spong reviewed (approved):
Checked out, tested locally, and code reviewed relevant detection-rules area files -- LGTM! 👍 🎉 

All previous issues I saw appear to be resolved @ersin-erdal -- I was able to receive notifications with each new setting, both configured together and independently. 🚀 

Only remaining nits I see are some styling issues with the KQL bar -- filter options button seems squished compared to the add filter button (see over diff in image), and as I think someone else mentioned the filter pills need some margin and are being stacked vertically instead of horizontally added:

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/233745987-46a679c9-1055-4f56-820a-419083c1c619.png" />
</p> 

Main KQL bar from Definition step:
<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/233745793-2ef8fdc7-1632-41f2-bdb3-45b1e4f21bc6.png" />
</p> 




Lastly, I have one question about the `23:59` fix w/ the ending time window:

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/233747385-28a4270c-d734-4e25-be2d-e57347f3aaa5.png" />
</p> 

Does this leave a 1min gap (23:59:00 -> 00:00:00) that an action wouldn't fire, or do you round up somewhere downstream to cover the full 24hrs?

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

ersin-erdal commented:
> Checked out, tested locally, and code reviewed relevant detection-rules area files -- LGTM! 👍 🎉
> 
> All previous issues I saw appear to be resolved @ersin-erdal -- I was able to receive notifications with each new setting, both configured together and independently. 🚀
> 
> Only remaining nits I see are some styling issues with the KQL bar -- filter options button seems squished compared to the add filter button (see over diff in image), and as I think someone else mentioned the filter pills need some margin and are being stacked vertically instead of horizontally added:
> 
> Main KQL bar from Definition step:
> 
> <img alt="" width="500" src="https://user-images.githubusercontent.com/2946766/233745793-2ef8fdc7-1632-41f2-bdb3-45b1e4f21bc6.png">
> 
> Lastly, I have one question about the `23:59` fix w/ the ending time window:
> 
> Does this leave a 1min gap (23:59:00 -> 00:00:00) that an action wouldn't fire, or do you round up somewhere downstream to cover the full 24hrs?

@spong 

There will be another PR for Conditonal Actions for `For each alert` type, i can fix that UI issue in that PR.

For `00:00 -> 24:00` issue i changed it to 23:59 to avoid `00:00 -> 00:00` (which means 00:00 -> 24:00) confusion.
If a user wants full day (24h) coverage they need to use a time range that has the same start and end times such as `00:00 -> 00:00`. or something like `15:00 -> 15:00`. If the end time is equal or lower than the start time, the framework considers the end time in the next day. But if you think that using `00:00 -> 00:00` is better we can simply switch to that.

spong commented:
> There will be another PR for Conditonal Actions for For each alert type, i can fix that UI issue (margin around the filter items) in that PR.

Sounds good 👍 

> For `00:00 -> 24:00` issue i changed it to 23:59 to avoid `00:00 -> 00:00` (which means 00:00 -> 24:00) confusion.
If a user wants full day (24h) coverage they need to use a time range that has the same start and end times such as `00:00 -> 00:00`. or something like `15:00 -> 15:00`. If the end time is equal or lower than the start time, the framework considers the end time in the next day. But if you think that using `00:00 -> 00:00` is better we can simply switch to that.

Thank you for clarifying the behavior here. 🙂 From a UX perspective, I think the UI should default the time-range to the full day, and users can modify from there. So if it's `00:00 -> 00:00` or `00:00 -> 24:00` I don't think it matters -- both read fine to me, but sounds like the former was causing that validation issue.

CoenWarmer commented:
This caused https://github.com/elastic/kibana/issues/155655, but we have a fix with https://github.com/elastic/kibana/pull/155796
